doc/developer/design/20230806_durable_catalog_state.md

```diff
@@ -0,0 +1,402 @@
+     * get_<object-type>_by_id(&self, id: I) -> ObjectType;
+     * As we separate the catalog state out from the rest of the
+     * Coordinator, we will most likely need to add these methods
+     * which should be fairly straight forward.
+     */
+
+    // ...
+
+    // Write
+
+    /// Creates a new durable catalog state transaction. 
+    fn transaction(&mut self) -> Transaction;
+
+    /// Commits a durable catalog state transaction.
+    fn commit_transaction(&mut self, tx: Transaction);
+
+    /*
+     * The majority of writes will go through a Transaction.
+     * However as an optimization it may be useful to have more
+     * targeted methods. We can add those on a case-by-case basis.  
+     */
+
+    /// Persist mapping from system objects to global IDs and fingerprints.
+    fn set_system_object_mapping(
+        &mut self,
+        mappings: Vec<SystemObjectMapping>,
+    );
+
+    // ...
+
+    // Misc
+
+    /// Confirms that this catalog is connected as the current leader.
+    ///
+    /// NB: We may remove this in later iterations of Pv2.
+    fn confirm_leadership(&mut self) -> Result<(), Error>;
+
+    /// Dumps the entire catalog contents in human readable JSON.
+    fn dump(&self) -> String;
+}
+```
+
+We will also add the following methods to initialize the durable catalog state structs. They do
+not open the catalog, but they may create connections over the network.
+
+```Rust
+fn stash_catalog_state(
+    now: NowFn,
+    stash_factory: StashFactory,
+    stash_url: String,
+    tls: MakeTlsConnector,
+) -> Box<dyn DurableCatalogState> {
+    // ...
+}
+
+fn persist_catalog_state(
+    now: NowFn,
+    shard_id: ShardId,
+    persist_client: PersistClient,
+) -> Box<dyn DurableCatalogState> {
+    // ...
+}
+
+fn shadow_catalog_state(
+    primary_state: Box<dyn DurableCatalogState>,
+    secondary_state: Box<dyn DurableCatalogState>,
+) -> Box<dyn DurableCatalogState> {
+    // ...
+}
+```
+
+These methods are needed to solve a bootstrapping problem. Some catalog state, such as deploy
+generation, is needed before attempting to open the catalog. So these methods will return unopened
+catalog states capable of returning only the state we need for bootstrapping.
+
+All current usages of `mz_adapter::catalog::storage::Connection` will be updated to
+use `Box<dyn DurableCatalogState>`. The `environmentd` binary will initialize
+the `Box<dyn DurableCatalogState>` using the methods above depending on command line arguments.
+
+### Stash implementation
+
+A stash based implementation will be fairly trivial to implement. This interface is already
+implemented using the stash, we will just need to change the name of some methods.
+
+### Persist Implementation
+
+This implementation is based
+on [Aljoscha's skunk works project](https://github.com/aljoscha/materialize/tree/skunk-log-for-coord-state).
+The entire state will be stored in a single persist shard as raw bytes with a tag to differentiate
```

Comment:
> Will this make reads very expensive because all reads will have to read over the entire stash? Because the stash includes storage usage, which can get into the millions of rows, that could be bad.

That's a good question, maybe this is motivation to move the storage usage out of the catalog and into it's own table/shard/storage-collection. Currently, it should only slow down bootstrapping because the persist contents will only be read during startup.

> Or is persist smart enough to mfp those reads down and only fetch the s3 blobs needed?

Hopefully this is the case, @danhhz do you know the answer to this?

> Does s3 compact things enough for any mfp benefit to go away (i.e., it compacts all shards into a single mega shard then a read still has to read lots of bytes)?

I don't think I fully understand this question. Which shards are you asking about?

Reply:
> ooof, yeah, storage usage definitely wants to be its own shard. i would have guessed that it lived in the storage controller, any idea why it ended up where it did?

I can't remember the definitive answer. I pulled up the original design doc: https://www.notion.so/materialize/Persist-Storage-Usage-Data-19bd434424df4d8395d9355e05f46b1a. It talks about wanting the source of truth to be a system table. Currently all system tables are backed by the stash which is probably why it landed there. 

In theory we could have some notion of append only system tables, that are not backed by the stash. We wouldn't retract and re-populate it's contents during bootstrap like other system tables. Storage usage and audit logs would benefit from this. The SQL logging work would also benefit from this, @umanwizard was discussing this last week, I'm not sure how he ended up resolving the issue. @umanwizard do not feel compelled to read anything in this PR other than this comment.

> if we pull out storage usage, what rough size do we expect the full catalog to be?

I'll try and grab some concrete numbers from the production sandbox for this.

Reply:
the storage usage really sounds a lot like a storage "introspection source", which is just a persist shard that storage maintains with information about what is going on within storage

the query log, I _think_, could potentially be modeled as a storage collection that is fed through a "MonotonicAppender" (which is how the first impl of webhook sources work)

Reply:
I think there may be some additional consistency constraints on this since it's used for billing. Storage usage collection timestamps are linearizable with system table writes. So it's impossible for the timestamp used in billing for some object to be before the timestamp when that object actually shows up in a system table.

I'm not sure how important that is, but I don't think it's possible with a storage introspection source/storage collection.

Reply:
It's important for any collection timestamp written to the `mz_storage_usage_by_shard` table to be greater than or equal to the current oracle timestamp

https://github.com/MaterializeInc/materialize/blob/29e3d2942f2148948c0dc5951d5d2f56e8a15346/src/adapter/src/coord/message_handler.rs#L164-L168

The best explanations for this are:

  * The commit message where this was introduced: https://github.com/MaterializeInc/materialize/commit/bd20ff91c59ec21d70aa2a933e30dadae9e2a0b9
  * The cloud issue that motivated this: https://github.com/MaterializeInc/cloud/issues/4942

I agree that a storage managed collection here would be ideal, but we'd need to somehow recover linearizability. The thesis way back when was "enforce real time recency on storage collections!" but that never got hooked up.

In a bit of a rush. Happy to explain further if any of the above is unclear.

Reply:
TIL! That's very useful context, thanks!

Looks like we lose control over the timestamp by the time things bubble up to the MonotonicAppender API, but specifying it is supported in the underlying `monotonic_append` call on the worker. Or this could also be a system table, which allows for specifying the timestamp.

Reply:
I'm somewhat leaning towards including the storage usage in the single shard, and making sure we come up with a solution to deal with it before getting rid of the stash backed catalog.

The storage usage data has three associated timestamps:

1. The timestamp used to write it into the stash (i.e. the [data, **ts**, 1] batch created by the stash).
2. The timestamp used to write it into the system table (i.e. the [data, **ts**, 1] batch created by persist).
3. The `collection_timestamp`, which is a column in the data itself (https://github.com/MaterializeInc/materialize/blob/60eeefe3203ffa9dacabc060dca3dfd5ef5cee8b/src/adapter/src/catalog/builtin.rs#L2068-L2071)

Currently, the only relationship between the three timestamps is that (2) is guaranteed to be larger than (3) and that no write/ddl operations happen at timestamps between (2) and (3). The linked issue was that (3) needs to be "greater than or equal to the current oracle timestamp". 

So I think we could safely pull the storage usage into a separate shard as long as we still use the timestamp oracle to generate `collection_timestamp`, maybe??

There's still some weird behavior with this table that I need to think about and can't quite make sense of. For example, 
```
SELECT * FROM mz_storage_usage_by_shard WHERE collection_timestamp <= ts AS OF ts
```
can return different results than
```
SELECT * FROM mz_storage_usage_by_shard WHERE collection_timestamp <= ts AS OF ts + n
```
s.t.  `n > 0`.

It feels like there may be some existing, but rare, issues with the table. It's complicated enough that it probably deserves it's own design doc/process.

Reply:
> I'm somewhat leaning towards including the storage usage in the single shard, and making sure we come up with a solution to deal with it before getting rid of the stash backed catalog.

wfm!